docs: order-management feature README + strip internal GUS refs @W-22821845 (#3907)
SalesforceCommerceCloud/pwa-kit: packages/template-retail-react-app/docs/README-ORDER-MANAGEMENT.md

# Order Management

The order detail page surfaces three [Salesforce Order Management

> (OMS/SOM)](https://help.salesforce.com/s/articleView?id=commerce.order_management.htm)

Reviewer: This link doesn't work - returns 404. I found this other doc: 
[Salesforce Order Management](https://help.salesforce.com/s/articleView?id=commerce.om_order_management.htm&type=5)
And
[Integrate Order Management with B2C Commerce](https://help.salesforce.com/s/articleView?id=commerce.om_impl_storefront_integration.htm&type=5)

shopper actions: **returning** eligible items, **cancelling** an order, and
**tracking** shipments. All three are driven by the OMS data attached to the order —
eligibility, returnable/cancellable quantities, statuses, and tracking details come
from OMS — and the authoritative accept/reject for the mutating actions happens
server-side via the corresponding order action.

There is **no feature flag**. Each action is gated entirely on data and shopper

> identity in [`order-detail.jsx`](../app/pages/account/order-detail.jsx). ECOM-only

Reviewer: ```suggestion
identity in [`order-detail.jsx`](../app/pages/account/order-detail.jsx). B2C Commerce-only
```

orders (no `omsData`) never expose the return or cancel flows, and fall back to raw
order/shipment status everywhere.

## Prerequisites

> These features require a **Salesforce Order Management (SOM) core org connected to

Reviewer: ```suggestion
These features require a Salesforce Order Management (SOM) core org connected to
```

the storefront's B2C Commerce instance**. SOM is what enriches orders with the
`omsData` this UI depends on — order- and item-level OMS data (returnable/cancellable

> quantities, item statuses, shipment tracking) and the `oms-return-order` /

Reviewer: It would be nice to provide links for those SCAPI order actions. I searched on the web and the doc portal but couldn't find the SCAPI endpoints for these: oms-return-order and oms-cancel-order.

> `oms-cancel-order` SCAPI order actions. See the [Order Management

Reviewer: ```suggestion
`oms-cancel-order` B2C Commerce API (SCAPI) order actions. See the [Order Management
```

> setup](https://help.salesforce.com/s/articleView?id=commerce.order_management.htm)

Reviewer: This link doesn't work. Replace with:
[Integrate Order Management with B2C Commerce](https://help.salesforce.com/s/articleView?id=commerce.om_impl_storefront_integration.htm&type=5)

docs for connecting and provisioning the org.

> Without a connected SOM org, orders carry no `omsData`: they are treated as ECOM-only,

Reviewer: ```suggestion
Without a connected SOM org, orders carry no `omsData`: they are treated as B2C Commerce-only,
```

the return and cancel actions never render, return reasons can't load, and the status

> badge falls back to the raw `order.status || order.omsData?.status`. Nothing errors — the features simply stay

Reviewer: ```suggestion
badge falls back to the raw `order.status || order.omsData?.status`. Nothing errors—the features simply stay
```

hidden. There is no storefront flag to turn them on; presence of OMS data on the order
is the switch.

## Order Returns

Registered shoppers can return eligible items. A shopper opens the **Return Items**
modal, picks items and per-item quantities, chooses a reason, reviews, and submits.
The order status badge then reflects the return's progress, including
partially-returned multi-unit lines.

### Eligibility

| Condition | Source | Meaning |
| --- | --- | --- |
| Registered shopper | `useCustomerType().isRegistered` | Guests never see the return UI. |
| Owns the order | `order.customerInfo.customerId === customerId` | A shopper can only return their own orders. |

> | OMS-managed order | `!!order.omsData` | ECOM-only orders carry no OMS data. |

Reviewer: ```suggestion
| OMS-managed order | `!!order.omsData` | B2C Commerce-only orders carry no OMS data. |
```

storefront change is needed.

### Error Handling

`classifyReturnError` inspects the HTTP status and (for `400`) the body's `errorCode`
discriminator, returning one of these kinds:

| `ReturnErrorKind` | Trigger | UI behavior |
| --- | --- | --- |
| `INVALID_REASON` | `400` `InvalidReasonCode` | Inline: reason no longer available, pick another. |
| `UNKNOWN_ITEMS` | `400` `UnknownProductItemIds` | Inline: affected items can't be returned. |
| `QUANTITY_EXCEEDED` | `400` `ReturnQuantityExceeded` | Inline: requested quantity too high. |
| `NOT_FOUND` | `404` | Terminal: order not found. |
| `CONFLICT` | `409` | Terminal: order no longer in a returnable state. |
| `NETWORK` | No HTTP response (fetch rejected) | Inline, retryable. |
| `UNKNOWN` | Any other status (incl. `400 OrderReturnFailed`, `5xx`) | Inline, retryable. |

The `409` case is the authoritative server-side refusal: even if the client thinks an
item is returnable, OMS has the final say. A mid-flow `401` is intercepted and
refreshed by the SDK auth layer before the classifier ever sees it.

## Order Cancellation

Registered shoppers can cancel an order that hasn't started fulfillment. The

> **Cancel order** button renders unconditionally (so its position is stable) but is

Reviewer: ```suggestion
**Cancel Order** button renders unconditionally (so its position is stable) but is
```

> disabled — with a screen-reader hint explaining why — when the order isn't eligible,

Reviewer: ```suggestion
disabled—with a screen-reader hint explaining why—when the order isn't eligible,
```

a cancellation just succeeded, or a terminal error made the order un-actionable.

### Eligibility

`canCancel` in [`order-detail.jsx`](../app/pages/account/order-detail.jsx) requires:

| Condition | Source |
| --- | --- |
| Registered shopper | `useCustomerType().isRegistered` |
| Owns the order | `order.customerInfo.customerId === customerId` |
| OMS-managed order | `!!order.omsData` |
| Every item fully cancellable | `item.omsData.quantityAvailableToCancel === item.omsData.quantityOrdered` for **all** `productItems` |

Cancellation is all-or-nothing: it's offered only when every line can still be

> cancelled in full. Once any unit has shipped, the order is no longer cancellable.

Reviewer: ```suggestion
cancelled in full. Once any unit has shipped, the order is no longer cancelable.
```

### How It Works

1. The shopper confirms in the
   [`CancelOrderModal`](../app/components/cancel-order-modal) (optionally supplying a
   reason).
2. The page submits via
   `useShopperOrdersMutation(ShopperOrdersMutations.CancelOmsOrder)`
   (`POST .../actions/oms-cancel-order`); the `reason` is sent only when provided.
3. On success an "Order cancelled" alert is shown (after a short delay so screen
   readers finish announcing the modal close) and the order status badge flips to

>    **Cancelled**.

Reviewer: ```suggestion
   **Canceled**.
```

4. On failure, cancellation follows the same convention as returns (see
   [Error Handling](#error-handling) above): a `404` or `409` is **terminal** — the

>    order can no longer be cancelled, so the button is permanently disabled with an

Reviewer: ```suggestion
   order can no longer be canceled, so the button is permanently disabled with an
```

   explanatory hint — while any other error shows a generic, retryable message. Unlike
   returns, cancellation has no per-`errorCode` classifier; it keys off the HTTP
   status alone.

Return feedback is kept separate from cancel feedback so the **Cancelled** badge
(which keys off the cancel result) never fires on a return success.

## Order Tracking

[`OrderTracking`](../app/components/order-tracking) is a presentational, bordered

> **tracking card** — one per shipment — rendered in a single flat **Tracking** section

Reviewer: Per our CX style, em dashes (—) should not be surrounded with spaces. I edited some of these but won't do that for all. Feel free to do a search and replace.
Search for: ' — '
Replace with: '—'

on the order detail page. Each card shows the carrier / shipping-method name, the
localized shipping status, the tracking number, and — when present — the expected and
actual delivery dates. A card carries **no shipping address** (addresses live with the
items; see [Shipments and addresses](#shipments-and-addresses) below).

The cards are built from `trackingEntries` (`order-detail.jsx`): one entry per

> `order.omsData.shipments[]`, falling back to `order.shipments[]` (ECOM) only when there

Reviewer: ```suggestion
`order.omsData.shipments[]`, falling back to `order.shipments[]` (B2C Commerce) only when there
```

> are no OMS shipments. ECOM-fallback cards have no provider, tracking URL, or dates

Reviewer: ```suggestion
are no OMS shipments. B2C Commerce-fallback cards have no provider, tracking URL, or dates
```

(those are OMS-only).

Key behaviors:

- **Carrier link safety.** When a tracking URL is available, the tracking number is
  hyperlinked through [`ensureExternalUrl`](../app/utils/url.js), which normalizes the
  scheme and rejects unsafe inputs (`javascript:` URLs, host-spoofing like
  `https://www.ups.com@evil.com`, and relative paths) — returning `undefined` so the
  number renders as plain text instead. Otherwise the tracking number is shown
  unlinked.
- **Date guarding.** Missing, null, or unparseable delivery dates render nothing
  rather than a misleading value. The `!value` guard is load-bearing: `new Date(null)`
  returns the epoch (1970-01-01), not an Invalid Date, so without it a null delivery
  date would display "31 Dec 1969".

> - **OMS-over-ECOM fallback.** The component receives already-resolved scalar props;

Reviewer: ECOM is only used internally. We should use "B2C Commerce" instead. I edited several instances of this. I suggest doing a search and replace to get them all.
Search for: 'ECOM'
Replace with: 'B2C Commerce'

  in `order-detail.jsx`. A provider-less OMS shipment borrows the delivery shipment's
  method name **only** in the unambiguous one-OMS-↔-one-delivery case; multi-shipment
  never joins, so a provider-less card simply shows no carrier name.

### Track Shipment action

A single **Track Shipment** order action sits in the actions row alongside Return
Items / Cancel order. Its source is `trackingUrlOptions` — the externalizable
(`ensureExternalUrl`'d) carrier URLs from `order.omsData.shipments` — so it can never
diverge from the tracking-number links in the cards. It has three states:

- **No externalizable URL** → the button renders **disabled** (kept visible, focusable
  via `aria-disabled` with a screen-reader hint "Tracking is not available for this
  order yet.").
- **Exactly one URL** → a single external-link button opening the carrier site in a new
  tab (the common single-shipment case).
- **Multiple URLs** → a dropdown (Popover) listing one carrier link per shipment, since
  a tracking entry can't be reliably tied to a specific shipment.

### Shipments and addresses

Shipment data arrives on the order in two lists with no correlation key between them:

> tracking info (status, tracking number/URL, dates) in `order.omsData.shipments`, and

Reviewer: ```suggestion
tracking info (status, tracking number/URL, and dates) in `order.omsData.shipments`, and
```
